# Botucatu_2024

New sequencing technologies, or NGS, have opened the door to massive sequencing of living organisms, offering the capability to sequence complex and contiguous genomes using very long reads. Among these technologies, nanopore sequencing stands out for its very low cost and its ever-increasing accuracy. A wide range of algorithms has been developed to analyze nanopore data. In this course, we will learn the methods and workflows for analyzing raw reads and achieving genome assembly. We will discover how to obtain nucleotide sequences from electrical signals, explore the most commonly used assembly paradigm for long fragments, test and compare several state-of-the-art assemblers, and finally evaluate the quality of the resulting assembly, with and without a reference genome.


# Course objectives:  
-To understand the advantages and disadvantages of Nanopore sequencing.   
-To learn about de novo genome assembly.  
-To gain practical experience with de novo genome assembly.  

# Program:
session_0: Before the start of the course, students should have installed the necessary software for the course, such as:  
-a virtual machine running Linux os (except for Linux and MacOs users).  
-[Anaconda](https://www.anaconda.com/products/distribution)   
-Students should be familiar with the [jupyter notebook](https://jupyter-notebook.readthedocs.io/en/stable/examples/Notebook/Notebook%20Basics.html) (type jupyter notebook in your teminal).  
Finaly you should download the program of the course usig git as follow: ->go to the git web site of the course ->clone the course: git clone 'adress'.
Students are invited to follow the session Intro_LINUX using jupyter notebook before the course

session_1: Introduction to Oxford Nanopore; Getting to know the `Fast5` format; Calling reads with `Guppy`.  
session_2: Quality control of fastq reads, Mapping reads against a reference genome with `Minimap2`.  
session_3: Genome assembly with `Minimap2` and `Miniasm`, `Flye` and `Shasta`.  
session_4: Polishing of the assemblies with `Racon` and `Medaka`.  
session_5: Quality of the assemblies with `Quast` and `BUSCO`.  

**Download data**.  
-Course materials will be made available on the day the course begins.
-Day 1: https://filesender.renater.fr/?s=download&token=04800581-a043-45f7-8692-9d82275d18f4  


Romain Guyot, IRD.  
[GenomeCoffea](https://www.genomecoffea.org)





Romain Guyot, IRD.
GenomeCoffea
